core/src/main/java/org/elasticsearch/indices/cluster/IndicesClusterStateService.java

```diff
@@ -227,7 +235,35 @@ private void applyDeletedIndices(final ClusterChangedEvent event) {
                 // First, though, verify the precondition for applying this case by
                 // asserting that the previous cluster state is not initialized/recovered.
                 assert previousState.blocks().hasGlobalBlock(GatewayService.STATE_NOT_RECOVERED_BLOCK);
-                indicesService.verifyIndexIsDeleted(index, event.state());
+                final IndexMetaData metaData = indicesService.verifyIndexIsDeleted(index, event.state());
+                if (metaData != null) {
+                    indexSettings = new IndexSettings(metaData, settings);
+                } else {
+                    indexSettings = null;
+                }
+            }
+            if (indexSettings != null) {
+                threadPool.generic().execute(new AbstractRunnable() {
```

Comment:
I know previous behavior was to spawn a thread every time. I think it will be cleaner to return a boolean from the deletion methods to say whether the index deletion was successful or is marked as pending and only spawn the thread for the latter.